File: baekjoon/python/17474 (수열과 쿼리 26).py
import sys

class Node:
    def __init__(self, max1, max2, max_cnt, s):
        self.max1 = max1
        self.max2 = max2
        self.max_cnt = max_cnt
        self.s = s

def compare_nodes(a, b):
    if a.max1 == b.max1:
        return Node(a.max1, max(a.max2, b.max2), a.max_cnt + b.max_cnt, a.s + b.s)
    
    if a.max1 > b.max1:
        a, b = b, a

    return Node(b.max1, max(a.max1, b.max2), b.max_cnt, a.s + b.s)

class SegmentTree:
    def __init__(self, arr):
        self.n = len(arr)
        self.arr = arr
        self.tree = [None] * (self.n * 4)
        self.lazy = [None] * (self.n * 4)
        self.build(0, self.n - 1, 1)

    def build(self, start, end, index):
        if start == end:
            self.tree[index] = Node(self.arr[start], -1, 1, self.arr[start])
            return
        
        mid = (start + end) >> 1
        self.build(start, mid, index << 1)
        self.build(mid + 1, end, (index << 1) | 1)
        self.tree[index] = compare_nodes(self.tree[index << 1], self.tree[(index << 1) | 1])

    def push(self, start, end, index):
        v = self.lazy[index]
        if v is None:
            return
        node = self.tree[index]
        if node.max1 <= v:
            self.lazy[index] = None
            return
        node.s -= node.max_cnt * (node.max1 - v)
        node.max1 = v
        if start != end:
            for child in (index << 1, (index << 1) | 1):
                if self.lazy[child] is None or self.lazy[child] > v:
                    self.lazy[child] = v if self.lazy[child] is None else min(self.lazy[child], v)
        self.lazy[index] = None

    def update(self, start, end, index, left, right, value):
        self.push(start, end, index)
        node = self.tree[index]
        if right < start or end < left or node.max1 <= value:
            return
        
        if left <= start and end <= right and node.max2 < value:
            self.lazy[index] = value
            self.push(start, end, index)
            return
        
        mid = (start + end) >> 1
        self.update(start, mid, index << 1, left, right, value)
        self.update(mid + 1, end, (index << 1) | 1, left, right, value)
        self.tree[index] = compare_nodes(self.tree[index << 1], self.tree[(index << 1) | 1])

    def query_max(self, start, end, index, left, right):
        self.push(start, end, index)
        if right < start or end < left:
            return float('-inf')
        
        if left <= start and end <= right:
            return self.tree[index].max1
        
        mid = (start + end) >> 1
        return max(self.query_max(start, mid, index << 1, left, right), self.query_max(mid + 1, end, (index << 1) | 1, left, right))

    def query_sum(self, start, end, index, left, right):
        self.push(start, end, index)
        if right < start or end < left:
            return 0
        
        if left <= start and end <= right:
            return self.tree[index].s
        
        mid = (start + end) >> 1
        return self.query_sum(start, mid, index << 1, left, right) + self.query_sum(start, end, (index << 1) | 1, left, right)

N = int(sys.stdin.readline())
A = list(map(int, sys.stdin.readline().split()))

seg_tree = SegmentTree(A)

M = int(sys.stdin.readline())
for _ in range(M):
    query = list(map(int, sys.stdin.readline().split()))
    if query[0] == 1:
        L, R, X = query[1:]
        seg_tree.update(0, N - 1, 1, L - 1, R - 1, X)

    elif query[0] == 2:
        L, R = query[1:]
        print(seg_tree.query_max(0, N - 1, 1, L - 1, R - 1))

    else:
        L, R = query[1:]
        print(seg_tree.query_sum(0, N - 1, 1, L - 1, R - 1))

File: baekjoon/c/17474 (수열과 쿼리 26).c
#include <stdio.h>
#include <stdlib.h>

#define INF 1000000001LL
typedef long long ll;

typedef struct {
    ll max1, max2, max_cnt, s;
} Node;

int N, M;
ll *A;
Node *tree;
ll *lazy;

ll maxll(ll a, ll b) { return a > b ? a : b; }
ll minll(ll a, ll b) { return a < b ? a : b; }

Node compare_nodes(Node a, Node b) {
    Node res;
    if (a.max1 == b.max1) {
        res.max1 = a.max1;
        res.max2 = maxll(a.max2, b.max2);
        res.max_cnt = a.max_cnt + b.max_cnt;
        res.s = a.s + b.s;
    } else {
        if (a.max1 > b.max1) {
            Node tmp = a; a = b; b = tmp;
        }
        res.max1 = b.max1;
        res.max2 = maxll(a.max1, b.max2);
        res.max_cnt = b.max_cnt;
        res.s = a.s + b.s;
    }
    return res;
}

void build(int start, int end, int idx) {
    lazy[idx] = -INF;
    if (start == end) {
        tree[idx].max1 = A[start];
        tree[idx].max2 = -INF;
        tree[idx].max_cnt = 1;
        tree[idx].s = A[start];
        return;
    }
    int mid = (start + end) / 2;
    build(start, mid, idx * 2);
    build(mid + 1, end, idx * 2 + 1);
    tree[idx] = compare_nodes(tree[idx * 2], tree[idx * 2 + 1]);
}

void push(int start, int end, int idx) {
    if (lazy[idx] == -INF) return;
    ll v = lazy[idx];
    Node *node = &tree[idx];
    if (node->max1 <= v) {
        lazy[idx] = -INF;
        return;
    }
    node->s -= node->max_cnt * (node->max1 - v);
    node->max1 = v;
    if (start != end) {
        for (int i = idx * 2; i <= idx * 2 + 1; ++i) {
            if (lazy[i] == -INF || lazy[i] > v) {
                lazy[i] = (lazy[i] == -INF) ? v : minll(lazy[i], v);
            }
        }
    }
    lazy[idx] = -INF;
}

void update_lazy(int start, int end, int idx) {
    push(start, end, idx);
}

ll interval_max(int start, int end, int idx, int left, int right) {
    update_lazy(start, end, idx);
    if (left > end || right < start) return -INF;
    if (left <= start && right >= end) return tree[idx].max1;
    int mid = (start + end) / 2;
    return maxll(
        interval_max(start, mid, idx * 2, left, right),
        interval_max(mid + 1, end, idx * 2 + 1, left, right)
    );
}

ll interval_sum(int start, int end, int idx, int left, int right) {
    update_lazy(start, end, idx);
    if (left > end || right < start) return 0;
    if (left <= start && right >= end) return tree[idx].s;
    int mid = (start + end) / 2;
    return interval_sum(start, mid, idx * 2, left, right) +
           interval_sum(mid + 1, end, idx * 2 + 1, left, right);
}

void update_tree(int start, int end, int idx, int left, int right, ll value) {
    update_lazy(start, end, idx);
    Node *node = &tree[idx];
    if (right < start || left > end || node->max1 <= value) return;
    if (left <= start && end <= right && node->max2 < value) {
        lazy[idx] = value;
        update_lazy(start, end, idx);
        return;
    }
    int mid = (start + end) / 2;
    update_tree(start, mid, idx * 2, left, right, value);
    update_tree(mid + 1, end, idx * 2 + 1, left, right, value);
    tree[idx] = compare_nodes(tree[idx * 2], tree[idx * 2 + 1]);
}

int main() {
    scanf("%d", &N);

    A = (ll*)malloc(sizeof(ll) * 4 * N);
    tree = (Node*)malloc(sizeof(Node) * (4 * N));
    lazy = (ll*)malloc(sizeof(ll) * (4 * N));

    for (int i = 0; i < N; ++i) scanf("%lld", &A[i]);
    build(0, N - 1, 1);

    scanf("%d", &M);
    for (int i = 0; i < M; ++i) {
        int q, L, R;
        ll X;
        scanf("%d", &q);
        if (q == 1) {
            scanf("%d %d %lld", &L, &R, &X);
            update_tree(0, N - 1, 1, L - 1, R - 1, X);
        } else if (q == 2) {
            scanf("%d %d", &L, &R);
            printf("%lld\n", interval_max(0, N - 1, 1, L - 1, R - 1));
        } else {
            scanf("%d %d", &L, &R);
            printf("%lld\n", interval_sum(0, N - 1, 1, L - 1, R - 1));
        }
    }

    free(A);
    free(tree);
    free(lazy);
    return 0;
}